### PR TITLE
SubSystem rewrite

### PR DESCRIPTION
**Pull request by ghost**
Misc:

+Fixes unreported issue with initializing lighting on a specific zlevel

+Fixes two similar issues with moveElement and moveRange. Where fromIndex or toIndex could be adjusted incorrectly in certain conditions. Potentially causing bad-sorts, or out of bound errors.

+Rewrites listclearnulls(list/L) to no longer iterate through L.len elements for every null in the list (plus 1). i.e. went from L.len*(number_of_nulls+1) list-element reads (best-case), to L.len list-element reads (worst-case)

+New proc/getElementByVar(list/L, varname, value) which finds the first datum in a list, with a variable named varname, which equals value. You can also feed it atoms instead of lists due to the way the in operator functions.

+Fixes an unreported issue with Yota's list2text rewrite. Under certain conditions, the first element would not be converted into a string. Causing type-mismatch runtimes.

+New global map_ready variable. This is not fully implemented yet, but will be used to avoid duplicate calls to initialize() for map objects.

+All turfs now maintain references to all lights currently illuminating them. This will mean higher memory use unfortunately, due to the huge number of turfs. However, it will speed up updateAffectingLights significantly. I've used list husbandry to reduce baseline memory usage, so it shouldn't be any worse than some past atmos modifications memory-wise.

-Removed 'quadratic lighting', can add this back at some point. Sorry.

+modified the way lum() works slightly, to allow turfs to have overridden delta-lumen. i.e. space cannot be illuminated more than its default ambiance. This allowed removal of some iffy special-snowflake lighting areas implemented by somebody else.

+Lighting images in the dmi can now use arbitrary naming schemes. It is reliant on order now. This allows the dmi to be replaced by simply dropping in a new dmi.

-Removed all subtypes of /area/shuttle. Shuttles now create duplicate 'rooms' of /area/shuttle. (More on this later). This will conflict with most maps. Guide on how to fix to follow.

+All verbs/tools relating to world.tick_lag were refactored to use world.fps. However old config text for setting tick_lag will still work (it converts the value to fps for you)

+MC stats improved using smoothing. They now have their own tab so they dont get in the way when you're playing as an admin.

-removed the push_mob_back stuff due to conflicting changes. Sorry Giacom.

_OK, NOW THE ACTUAL INTERESTING STUFF_

Following systems moved over to subsystem datums:
air_master
garbage_manager
lighting_controller
process_mobs (aka Life())
nanomanager
power
sun
pipenets
AFK kick loops
shuttle_controller (aka emergency shuttle/pods), supply_shuttle and other shuttles
voting
bots
radio
diseases
events
jobs
objects
ticker

Subsystems hooks and variables should be commented fairly in-depth. If anything isn't particularly clear, please make an issue.

Many system-specific global variables have been refactored into

All tickers which previously used world.timeofday now use world.time

some subsystems can iterate before round start. this resolves the issue with votes not working pregame

NOTE: info regarding shuttles rewrite is missing as that can only be finalised after the map is updated. Which I will do on 1st/2nd Jan.

### CODE REVIEW

**Comment by paprka** [comment]
Carn's working on the map now as well as the conflicts.

File: code/controllers/_DynamicAreaLighting_TG.dm [hunk 1]
```diff
@@ -306,20 +264,27 @@
...
-			A.light.changed = 1			//force it to update at next process()
+
+/atom/movable/UpdateAffectingLights()
+	if(isturf(loc))
+		loc.UpdateAffectingLights()
+
+/turf/UpdateAffectingLights()
+	if(affecting_lights)
+		for(var/thing in affecting_lights)
+			thing:changed = 1			//force it to update at next process()
```
Comment (Razharas) [inline code/controllers/_DynamicAreaLighting_TG.dm]: Either typecast (without checking if you are concerned with speed) or at least rename the var into something describing what exactly is the things that is supposed the be updated in next process()
I know that affecting_lights is supposed to only hold special things, but its for better understandability of code

**Comment by YotaXP** [comment]
Dag, you've been busy.  Can someone compile a list of related issues?

File: code/controllers/_DynamicAreaLighting_TG.dm (hunk 1 above)
Comment (Carn) [inline code/controllers/_DynamicAreaLighting_TG.dm]: no

**Comment by Carn** [comment]
clarify: >list of related issues?

**Comment by CorruptComputer** [comment]
Probably the issues that this fixes.

**Comment by Carn** [comment]
well duh
I mean why

**Comment by YotaXP** [comment]
So we can close them afterward.

**Comment by paprka** [comment]
Yeah the maintainers (the good ones anyway :VVVVVVVVV) regularly go through the issues and cut down ones that are already fixed, but htis probably fixes a metric fuckton of shit.

**Comment by Carn** [comment]
But I can just add them to the PR =p

**Comment by paprka** [comment]
Why the fuck is asteroid still in travis goddamn

**Comment by Iamgoofball** [comment]
you need to run mapmerge

**Comment by Carn** [comment]
The map is frozen, what's the point?

**Comment by paprka** [comment]
Map is frozen and even if she did run map merge it would conflict with literally any other map edit due to the number of line changes despite that

**Comment by CorruptComputer** [comment]
Was just playing to see if performance improvements and I found a couple bugs. On the lobby tab it always says "Players:  1" and it doesn't list the ckeys.

**Comment by Iamgoofball** [comment]
Yes but if you at least make some effort to run mapmerge people can somewhat salvage some pull requests that change the map

**Comment by Carn** [comment]
@corruptcomputer ckeys was removed. Player count should work

**Comment by Giacom** [comment]
>-removed the push_mob_back stuff due to conflicting changes. Sorry Giacom.

Oh well, sacrifices have to be made. :hocho: :snowflake:

**Comment by Carn** [comment]
sorry giacom, it might be able to come back at some point.

**Comment by Giacom** [comment]
It's alright, I'm not bothered.

**Comment by Carn** [comment]
I'm stuck on a bug where atmos machinary doesn't seem to work. I think it's a merge error somewhere

**Comment by Carn** [comment]
Should be ready for a baptism of fire

**Comment by pudl** [comment]
carn, can you please write up an explanation for how the new shuttle shit works?

**Comment by Carn** [comment]
'carn, can you please write up an explanation for how the new shuttle shit works?'
On my todo, I plan on simplifying it a bit at some point anyway so don't worry too much.

Yota, yes that's what they are referring to. I will sort it out, there's better ways of doing than some dodgy math anyway

**Pull request merged**